src/board.rs

```diff
@@ -198,6 +205,75 @@ impl Board {
     }
 }
 
+// Errors are raised if the FEN string is invalid
+// i.e., different that pbnrqk
+fn char_to_piece(piece: &str, color: Color) -> Result<Piece, ParseFenError> {
+    match piece {
+        "p" => Ok(Piece {
+            piece_type: PieceType::Pawn,
+            color,
+        }),
+        "b" => Ok(Piece {
+            piece_type: PieceType::Bishop,
+            color,
+        }),
+        "n" => Ok(Piece {
+            piece_type: PieceType::Knight,
+            color,
+        }),
+        "r" => Ok(Piece {
+            piece_type: PieceType::Rook,
+            color,
+        }),
+        "q" => Ok(Piece {
+            piece_type: PieceType::Queen,
+            color,
+        }),
+        "k" => Ok(Piece {
+            piece_type: PieceType::King,
+            color,
+        }),
+        _ => Err(ParseFenError::InvalidPiecePlacement(piece.to_string())),
+    }
+}
+
+// Errors are raised if the FEN string is invalid
+// char_to_piece is responsible for raising the error
+impl FromStr for Board {
+    type Err = ParseFenError;
+
+    fn from_str(piece_placement: &str) -> Result<Self, Self::Err> {
+        let mut board = Board::new_empty_board();
+
+        let mut rank = 7;
+        let mut file = 0;
+
+        for c in piece_placement.chars() {
+            match c {
+                '0'..='8' => {
+                    let empty_squares = c.to_digit(10).unwrap() as usize;
+                    file += empty_squares;
+                }
+                '/' => {
+                    rank -= 1;
+                    file = 0;
+                }
+                'a'..='z' => {
+                    let new_piece = char_to_piece(&c.to_lowercase().to_string(), Color::Black)?;
+                    board.set_piece((rank * BOARD_SIZE + file).try_into().unwrap(), new_piece);
+                    file += 1;
+                }
+                'A'..='Z' => {
+                    let new_piece = char_to_piece(&c.to_lowercase().to_string(), Color::White)?;
+                    board.set_piece((rank * BOARD_SIZE + file).try_into().unwrap(), new_piece);
+                    file += 1;
+                }
+                _ => break,
+            }
+        }
+        Ok(board)
+    }
+}
 
 #[cfg(test)]
 mod tests {
```

Comment:
if board_tests exists, lets move these tests there

Reply:
(if these tests are still needed)

Reply:
For now they can remain although they will be replaced with what you have in your branch for pawn/knight moves. However,  I agree. I ll move them to the board_tests file